8271471: [IR Framework] Rare occurrence of "<!-- safepoint while printing -->" in PrintIdeal/PrintOptoAssembly can let tests fail

chhagedorn opened the pull request:
A test VM used by the IR framework sometimes prints `<!-- safepoint while printing -->` in the middle of emitting a `PrintIdeal` or `PrintOptoAssembly` output which could lead to IR matching failures:
https://github.com/openjdk/jdk/blob/489e5fd12a37a45f4f5ea64b05f85c6f99f70811/src/hotspot/share/utilities/ostream.cpp#L918-L927

I thought about just bailing out of IR matching if this string is found after a failure but this issue also affects internal framework tests (I observed one case locally where this happened in the test `TestIRMatching`, letting it fail).

Handling `<!-- safepoint while printing -->` makes things more complicated for the IR framework tests. I'm not sure about the value of printing this message in the first place. But if nobody objects, I suggest to either remove it or at least guard it with `Verbose`, for example. I went with the latter solution in this PR.

Thanks,
Christian

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271471](https://bugs.openjdk.java.net/browse/JDK-8271471): [IR Framework] Rare occurrence of "<!-- safepoint while printing -->" in PrintIdeal/PrintOptoAssembly can let tests fail


### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4932/head:pull/4932` \
`$ git checkout pull/4932`

Update a local copy of the PR: \
`$ git checkout pull/4932` \
`$ git pull https://git.openjdk.java.net/jdk pull/4932/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4932`

View PR using the GUI difftool: \
`$ git pr show -t 4932`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4932.diff">https://git.openjdk.java.net/jdk/pull/4932.diff</a>

</details>

chhagedorn commented:
/label add hotspot-compiler

vnkozlov reviewed (approved):
Agree.

dholmes-ora commented:
I think the expectation is that a safepoint whilst printing and holding the tty lock should be a very rare thing, possibly indicative of an error, and so something we want to know about. If this is hidden by Verbose then it will effectively always be hidden and we won't spot this.

Do you understand why this is getting printed in this context, and that it doesn't indicate a problem?

Thanks,
David

chhagedorn commented:
I see your concern. Let me try to reproduce it and get some more information when it happens again. This might take some time. I'll get back to you.

chhagedorn commented:
I played around with it and interestingly this code path is hit quite often when just running `java -Xbatch -XX:+PrintIdeal -XX:+PrintOptoAssembly --version`.

I had a closer look at one of these occurrences. In this case, the reason we are executing `break_tty_lock_for_safepoint()` with a tty lock (i.e. at a safepoint while printing) was the following:
At some point, `SharedRuntime::handle_wrong_method_ic_miss` is called in thread X due to having cached the wrong method. This code then creates a new inline cache stub. In the meantime, the VM thread Y is waiting for a next VM operation in the loop. When the VM thread is calling `VMThread::setup_periodic_safepoint_if_needed()` it notices that a cleanup is needed (`SafepointSynchronize::is_cleanup_needed()` is true) due to the new IC stub. The cleanup operation is chosen as next VM operation. To execute this operation, it requires a safepoint. It sets the polling word of all threads, including the C2 compiler thread and `SafepointMechanism::local_poll_armed()` is now true. Unfortunately, the C2 compiler thread is currently printing with a tty lock acquired either at:
https://github.com/openjdk/jdk/blob/eec64f55870cf51746755d8fa59098a82109e826/src/hotspot/share/opto/output.cpp#L1870-L1871
or at
https://github.com/openjdk/jdk/blob/eec64f55870cf51746755d8fa59098a82109e826/src/hotspot/share/opto/compile.cpp#L789-L790
Therefore, we break the lock inside `ThreadInVMfromNative()` at the next print call in the C2 compiler thread because `SafepointMechanism::local_poll_armed()` is true.

To me this seems okay and not a problem but just bad luck due to timing.

dholmes-ora commented:
Thanks @chhagedorn . I can see that this message is somewhat expected in this context. Is it really that difficult to ignore in the IR matching? This is XML-ouput and it is formatted as a comment, so shouldn't the matching logic ignore XML comments?

chhagedorn commented:
I can think of a few options:
1. Keep matching code but if we catch an `IRViolationException` we additionally check if this tty message was found on the matched output. If that is the case, we bail out without an exception:
  - +: Easy and works well for normal user written IR tests. We only need to do this tty if there is actually an IR failure being reported. I'd say that it is unlikely that an IR test is written and is not executed a single time successfully without a bailout before integrating it.
  - -: Does not work well with framework internal tests which sometimes expect an exception (to check a few things on it) and sometimes not. Each test somehow needs to check if there was a bailout to reliably work. I guess it is okay to just ignore it for the "no exception expected" cases. We would catch possible errors in the framework the next time. For the "exception expected" tests, it's a little more difficult. The most simple thing to do is probably to just repeat a single test or a group of tests in an internal test file.
2. Change the regex matching code to ignore the tty message:
  - +: No changes to internal tests.
  - -: Given that it occurs very rarely, it has quite an overhead.
3. Strip a tty message when reading the `PrintIdeal`/`PrintOptoAssembly` entries from the log file when setting up the output to be parsed:
  - +: No changes to internal tests.
  - -: Given that it occurs very rarely, it has some overhead but probably less than the second option. We are reading each line anyways (I'm not actually parsing XML).

In terms of performance, it is probably best to bail out and do a one time effort to make the internal framework tests aware of this bailout scenario (option 1). While thinking about it for a second time, it might not be that complicated as I thought before. My original idea was just to first propose a removal of this tty message before going into fixing this rare case. But of course, given that there are concerns, I'm considering one of these options (my favorite is option 1).

TobiHartmann commented:
I would also go with option 1). For the framework internal tests that expect and exception, couldn't you simply add something like a `hitSafepointWhilePrinting()` method to `IRViolationException` that you can use to check if verification of the exception should be skipped?

chhagedorn commented:
> I would also go with option 1). For the framework internal tests that expect and exception, couldn't you simply add something like a `hitSafepointWhilePrinting()` method to `IRViolationException` that you can use to check if verification of the exception should be skipped?

Thanks Tobias, I therefore went with option 1 and reverted the HotSpot change. The problem is when no `IRViolationException` is thrown due to the bailout in the `IRMatcher`. I made the internal framework tests aware of it by replacing "shouldNotReach" checks by `shouldHaveCaughtException` methods which check if we had this `tty` message somewhere. If not then the test fails. Normal user written IR tests just pass when an `IRViolationException` would have been thrown but the matched input contained the `tty` message.

TobiHartmann commented on test/hotspot/jtreg/testlibrary_tests/ir_framework/tests/Utils.java:
```diff
@@ -0,0 +1,65 @@
+import compiler.lib.ir_framework.driver.TestVMProcess;
+
+public class Utils {
+    public static void shouldHaveCaughtException() {
```
Should these methods be renamed to `shouldHaveThrownException`? Because that method could also be used in a context where we don't catch.

chhagedorn replied:
Agreed, that's a better name!

TobiHartmann commented on test/hotspot/jtreg/testlibrary_tests/ir_framework/tests/Utils.java:
```diff
@@ -0,0 +1,65 @@
+        if (!hitSafePointWhilePrinting) {
+            throw new ShouldHaveCaughtException();
+        }
+    }
```
Why don't we simply do:
```
    for (Scenario s : scenarios) {
        shouldHaveCaughtException(s.getTestVMOutput());
    }
```

chhagedorn replied:
I changed that code, see comment below.

chhagedorn commented:
I did a renaming and changed the way `TestScenarios` throws exceptions (was wrong how I had it before). `TestScenarios` should actually do some more finegrained checks and depending on which scenarios bailed out do different things. I added some comments for them.

TobiHartmann commented on test/hotspot/jtreg/testlibrary_tests/ir_framework/tests/TestScenarios.java:
```diff
@@ -44,15 +44,28 @@ public static void main(String[] args) {
+                // Not all scenarios had a bailout which means that at least one exception should have been thrown.
+                Asserts.fail("Should have thrown an exception");
+            }
+
```
Newline can be removed

TobiHartmann reviewed (approved):
Looks good to me. I'm just wondering if we need all that `notAllBailedOut`/`noneBailedOut` complexity for handling scenarios. Given that safepoints during printing are rare, couldn't we simply skip verification of the exception message if one scenario hit a bailout?

chhagedorn commented:
> Looks good to me. I'm just wondering if we need all that `notAllBailedOut`/`noneBailedOut` complexity for handling scenarios. 

I think we need the `notAllBailedOut()` function to verify that there was not a scenario without a bailout (i.e. missed to throw an exception). The code of `notAllBailedOut()` could be simplified to use `anyMatch` instead.

> Given that safepoints during printing are rare, couldn't we simply skip verification of the exception message if one scenario hit a bailout?

Yes, that's probably easier. We can also first check if the expected message was even mismatched in `TestScenarios` before actually doing the `tty` message check. I changed that by turning `noneBailedOut` into `anyBailedOut` which uses `anyMatch`.

But yes, I agree that it's some complexity for a rather rare case. I pushed an updated with these changes.

TobiHartmann reviewed (approved):
Looks good, thanks for making these changes!

chhagedorn commented:
Thank you Tobias for your careful review!

vnkozlov reviewed (approved):
Good

chhagedorn commented:
Thanks Vladimir for your review!

chhagedorn commented:
/integrate
